data_steward/cdr_cleaner/cleaning_rules/truncate_fitbit_data.py

```diff
@@ -4,51 +4,44 @@
 Original Issue: DC-1046
 
 The intent is to ensure there is no data after the cutoff date for participants in
-Activity Summary, Heart Rate Minute Level, Heart Rate Summary, and Steps Intraday tables
-by sandboxing the applicable records and then dropping them.
+the fitbit tables by sandboxing the applicable records and then dropping them.
 """
 
 # Python Imports
 import logging
 from datetime import datetime
 
 # Project Imports
-import common
-import constants.cdr_cleaner.clean_cdr as cdr_consts
+from common import JINJA_ENV, FITBIT_TABLES
+from resources import validate_date_string, fields_for
+from constants.cdr_cleaner.clean_cdr import FITBIT, DESTINATION_TABLE, DESTINATION_DATASET, QUERY, DISPOSITION
 from cdr_cleaner.cleaning_rules.base_cleaning_rule import BaseCleaningRule
 from constants.bq_utils import WRITE_TRUNCATE
 
 LOGGER = logging.getLogger(__name__)
 
-FITBIT_DATE_TABLES = [
-    common.ACTIVITY_SUMMARY, common.HEART_RATE_SUMMARY, common.SLEEP_LEVEL,
-    common.SLEEP_DAILY_SUMMARY
-]
-FITBIT_DATETIME_TABLES = [common.HEART_RATE_MINUTE_LEVEL, common.STEPS_INTRADAY]
-
-FITBIT_TABLES_DATE_FIELDS = {
-    common.ACTIVITY_SUMMARY: 'date',
-    common.HEART_RATE_SUMMARY: 'date',
-    common.SLEEP_DAILY_SUMMARY: 'sleep_date',
-    common.SLEEP_LEVEL: 'sleep_date'
-}
-FITBIT_TABLES_DATETIME_FIELDS = {
-    common.HEART_RATE_MINUTE_LEVEL: 'datetime',
-    common.STEPS_INTRADAY: 'datetime',
-}
-
 # Save rows that will be dropped to a sandboxed dataset
-SANDBOX_QUERY = common.JINJA_ENV.from_string("""
-CREATE OR REPLACE TABLE `{{project}}.{{sandbox}}.{{intermediary_table}}` AS (
+SANDBOX_QUERY = JINJA_ENV.from_string("""
+CREATE OR REPLACE TABLE `{{project_id}}.{{sandbox_id}}.{{intermediary_table}}` AS (
 SELECT * 
-FROM `{{project}}.{{dataset}}.{{table_name}}`
-WHERE {{date_field}} > {{cutoff_date}})""")
+FROM `{{project_id}}.{{dataset_id}}.{{fitbit_table}}`
+{% if datetime_fields and date_fields %}
+  WHERE (GREATEST({{date_fields}}) > DATE("{{truncation_date}}"))
```

Comment:
I think the `COALESCE()` logic should be in this JINJA template rather than in `get_query_specs()` so we can see the whole query structure in one place. 

Can we change this 
`  WHERE (GREATEST({{date_fields}}) > DATE("{{truncation_date}}"))`
to be something like this?
```
# date_fields is the list of field names with datatype==date
# do the same for datetime_fields

  WHERE (GREATEST(
    {% for field in date_fields %}
    COALESCE({{field}}, DATE("1900-01-01"))
    {% endfor %}{% if not loop.last -%}, {% endif %}
  ) > DATE("{{truncation_date}}"))`
```
Or if `GREATEST` can have both date and datetime values at once, we can do it like this and we can remove the `if/elif` conditions in the JINJA template:
```
  WHERE (GREATEST(
    {% for field in date_fields + datetime_fields %}
    COALESCE({{field}}, DATE("1900-01-01"))
    {% endfor %}{% if not loop.last -%}, {% endif %}
  ) > DATE("{{truncation_date}}"))`
```
^ It's mock - some syntax might be wrong.

Reply:
Great idea to use `GREATEST()` here, by the way!

Reply:
I like your idea of combining the fields!!! 

I went ahead and combined them in the `get_query_specs()` to simplify things. After doing that I realized we might not need a loop in the query at all which I find more readable, but this would require the COALESCE statement to stay in the `get_query_specs()`.  As readability is in the eye of the beholder, what do you think is best?

Reply:
Your latest update is the most straightforward and most readable. You are correct that doing everything in the JINJA template here could make it look messier. Thanks for the update!